### PR TITLE
ENH: Method for renaming variables

### PR DESCRIPTION
**Pull request by rstoneback**
# Description

Addresses #91 

Includes a .rename variable method attached to the Instrument object. Renames variables for both data and metadata, keeping both objects in sync. Includes support for changing the name of variables in higher-order datasets forced into pandas. Examples are included in the docstring.

This feature could be sneakily snaked into v2.2 to support changing ICON variables to something easier to work with.

## Type of change

Please delete options that are not relevant.

- New feature (non-breaking change which adds functionality)

# How Has This Been Tested?

 - Developed new tests included in test_instrument
 - Ran pytests locally

# Checklist:

- [x] Make sure you are merging into the ``develop`` (not ``master``) branch
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules
- [x] Add a note to ``CHANGELOG.md``, summarizing the changes

### CODE REVIEW

File: pysat/_instrument.py [hunk 1]
```diff
@@ -619,6 +619,116 @@ def __setitem__(self, key, new):
...
+
+        Parameters
+        ----------
+        names : dict or other map
+            Existing names are keys, values are new names
+
+        Returns
+        -------
+        Void
+            Object modified in place.
```
Comment (aburrell) [inline pysat/_instrument.py:633]: Not needed.
```suggestion
```

File: pysat/tests/test_instrument.py [hunk 2]
```diff
@@ -615,6 +615,57 @@ def test_getting_all_data_by_numpy_array_of_int(self):
...
+    # --------------------------------------------------------------------------
+    def test_basic_variable_renaming(self):
+        # test single variable
+        self.testInst.load(2009, 1)
+        self.testInst.rename({'uts': 'uts1'})
+        assert 'uts1' in self.testInst.data
+        assert 'uts1' in self.testInst.meta
+        assert 'uts' not in self.testInst.data
+        assert 'uts' not in self.testInst.meta
+
```
Comment (aburrell) [inline pysat/tests/test_instrument.py:631]: This is one test

File: pysat/tests/test_instrument.py [hunk 3]
```diff
@@ -615,6 +615,57 @@ def test_getting_all_data_by_numpy_array_of_int(self):
...
+        assert 'uts' not in self.testInst.data
+        assert 'uts' not in self.testInst.meta
+        assert 'uts1' not in self.testInst.data
+        assert 'uts1' not in self.testInst.meta
+
+        assert 'mlt2' in self.testInst.data
+        assert 'mlt2' in self.testInst.meta
+        assert 'mlt' not in self.testInst.data
+        assert 'mlt' not in self.testInst.meta
+
```
Comment (aburrell) [inline pysat/tests/test_instrument.py:646]: This is another test

File: pysat/tests/test_instrument.py [hunk 4]
```diff
@@ -615,6 +615,57 @@ def test_getting_all_data_by_numpy_array_of_int(self):
...
+        # check for error for unknown variable name
+        with pytest.raises(ValueError):
+            self.testInst.rename({'help': 'I need somebody'})
+
```
Comment (aburrell) [inline pysat/tests/test_instrument.py:650]: This is a third test

File: pysat/tests/test_instrument.py [hunk 5]
```diff
@@ -615,6 +615,57 @@ def test_getting_all_data_by_numpy_array_of_int(self):
...
+    def test_ho_pandas_variable_renaming(self):
```
Comment (aburrell) [inline pysat/tests/test_instrument.py:651]: This test can probably also be broken up

File: pysat/tests/test_instrument.py [hunk 6]
```diff
@@ -615,6 +615,57 @@ def test_getting_all_data_by_numpy_array_of_int(self):
         else:
             assert a.sizes['time'] == 5
 
+    # --------------------------------------------------------------------------
+    #
+    # Test variable renaming
```
Comment (aburrell) [inline pysat/tests/test_instrument.py:620]: The first two tests I identify below can be combined into a single parametrized test.

Comment (rstoneback) [reply pysat/tests/test_instrument.py:620]: Thanks for the review @aburrell. Changes are in.

File: pysat/tests/test_instrument.py [hunk 7]
```diff
@@ -615,6 +615,72 @@ def test_getting_all_data_by_numpy_array_of_int(self):
...
+            if self.testInst.name == 'testing2D':
+                self.testInst.load(2009, 1)
+                # check for error for unknown HO variable name
+                with pytest.raises(ValueError):
+                    sub_dict = {'help': 'I need somebody'}
+                    self.testInst.rename({'profiles': sub_dict})
+                # check for bad column name when doing HO renaming
+                with pytest.raises(ValueError):
+                    sub_dict = {'help': 'I need somebody'}
+                    self.testInst.rename({'fake_profiles': sub_dict})
```
Comment (aburrell) [inline pysat/tests/test_instrument.py:682]: This test can also be parametrized.

Comment (rstoneback) [reply pysat/tests/test_instrument.py]: Updated. Parameters for all! Tests pass locally.

**Comment by rstoneback** [comment]
Now for the double sneaky part.

The original method this came from had a feature whereby the new variable names in data would be lower case, however the original case is stored in Meta. This makes it easier on users writing code, but, when the file is written the preserved case in Meta is used. Fancy names in file, easier names in python when working data.

It is a very simple change code wise, probably requires more documentation than anything. Any thoughts/feelings on including that here?

I can re-use the testing to a large extent since it is all parameterized now :)

**Comment by aburrell** [comment]
> I can re-use the testing to a large extent since it is all parameterized now :)

Yay!!!

**Comment by aburrell** [comment]
>  Any thoughts/feelings on including that here?

Probably, but reading your statement over a few times and I am not sure I know what you're getting at.

**Comment by rstoneback** [comment]
The method should pop up in the API section of the documentation. Can you provide a bit more about the documentation updates you are looking for?

**Comment by aburrell** [comment]
As a new feature, I'd expect it to be discussed in whichever part of the documentation talks about how to use that type of thing and an example added to the examples.  For this one in particular, adding to the discussion on creating a new instrument and how this function can help may be the right place.

**Comment by rstoneback** [comment]
> > Any thoughts/feelings on including that here?
> 
> Probably, but reading your statement over a few times and I am not sure I know what you're getting at.

Fair enough. I can add a flag to rename that would do the following.
 - Force all variable name provided by user to lower case when naming things in inst.data
 - When renaming the names in inst.meta that case provided by the user will be retained (note variable access is case insensitive for meta)
 - The names in inst.meta are used when writing netCDF4 files.
 - A user can work with data using, say, inst['variable_name'], but when they write the file it would come out as, say, 'Variable_Name' or whatever other case was originally provided as part of rename.
- TL;DR Trying to make it easier on users to work with data while simultaneously making it easier to make files that pass standards that aren't user friendly

**Comment by rstoneback** [comment]
Cool. I'll work on it after I make a pass through all the other pysat activity :)

Adding this feature also makes the rename more complicated than standard so I agree it would be worth spending some additional time in the documentation on it.

**Comment by rstoneback** [comment]
The new feature is in and I've expanded the examples section to include using the new lowercase_data_labels feature. I also scanned the documentation and I did not see a very good place to put in rename. I did think of either a new section or new example - using pysat to make files for archive at NASA's SPDF/CDAWeb consistent with the latest explorers. This will tie in a few features in pysat, while in the docstrings, aren't in the additional custom written documentation. I think it will be good but out of scope for this particular pull.

**Comment by rstoneback** [comment]
See issue #463

**Comment by rstoneback** [comment]
> While I disagree in principle, having the issue in and expanding the examples are good enough for our current standards.

We should talk about this at the next pysat meeting and sort out what our standards are going to be. 
Thanks for the review feedback! Always appreciated.

**Comment by rstoneback** [comment]
For reasons unknown some lines that should absolutely be tested aren't showing up as tested in coveralls. Going to sort out why.

**Comment by jklenzing** [comment]
Looking through the coverage of test_instrument, I don't see a class that invokes the 2d testing instrument, only 1d.

**Comment by jklenzing** [comment]
Somehow the `import logging` line got removed from one of the tests in the merged version.  Adding it back.

**Comment by jklenzing** [comment]
@aburrell, the coverage issues should be sorted out.  Is this still "approved" to go?

**Comment by rstoneback** [comment]
Nice. Thanks everyone. Merging :)
